torch_geometric/loader/neighbor_loader.py

```diff
@@ -196,19 +204,28 @@ def _hetero_sparse_neighbor_sample(self, index_dict: Dict[str, Tensor],
                 self.replace,
                 self.directed,
             )
+
         return node_dict, row_dict, col_dict, edge_dict
 
-    def __call__(self, index: Union[List[int], Tensor]):
-        if not isinstance(index, torch.LongTensor):
-            index = torch.LongTensor(index)
+    def __call__(self, index: Union[List[int], Tensor, HeteroNodeList]):
+        if self.data_cls == 'custom' or issubclass(self.data_cls, HeteroData):
+
+            if isinstance(index, List) and isinstance(index[0], Tuple):
+                index_dict = from_node_list(index)
+            else:
+                index = torch.LongTensor(index)
+                index_dict = {self.input_type: index}
+
+            batch_sizes = {
+                node_type: index.numel()
+                for node_type, index in index_dict.items()
+            }
 
-        if self.data_cls != 'custom' and issubclass(self.data_cls, Data):
-            return self._sparse_neighbor_sample(index) + (index.numel(), )
+            return self._hetero_sparse_neighbor_sample(index_dict) + (
+                batch_sizes, )
```

Comment:
This makes sense, and you are right that it somewhat conflicts with the interface in #5312. I think this is totally okay; that interface will likely change significantly over the coming week or two (we also need to support link-level neighbor sampling, etc.).

If it is alright with you, I would propose first merging 5312, and then adapting that interface as part of this PR to support returning a dict of batch sizes for each node type. Wdyt?

Reply:
Agreed! I see its merged now so let me rethink this a little based on what you've added.